modules/service/src/main/scala/lucuma/itc/service/ItcMappings.scala

```diff
@@ -83,272 +52,216 @@ object ItcMapping extends ItcCacheOrRemote with Version {
-                                  BuildInfo.ocslibHash,
-                                  ccds,
-                                  charts.flatMap(_.charts),
-                                  peakFinalSNRatio,
-                                  atWvFinalSNRatio,
-                                  peakSingleSNRatio,
-                                  atWvSingleSNRatio
-          )
-        )
-      }
-      .recover(recoverResult)
+  )(asterismRequest: AsterismGraphRequest): F[Result[SpectroscopyGraphResult]] =
+    asterismRequest.toTargetRequests
+      .parTraverse: (targetRequest: TargetGraphRequest) =>
+        graphFromCacheOrRemote(targetRequest)(itc, redis).attempt
+          .map: (result: Either[Throwable, GraphResult]) =>
+            TargetGraphOutcome:
+              result.bimap(buildError, buildTargetGraphResult(asterismRequest.significantFigures))
+      .map: (targetGraphs: NonEmptyChain[TargetGraphOutcome]) =>
+        SpectroscopyGraphResult(
+          version(environment).value,
+          BuildInfo.ocslibHash,
+          AsterismTargetGraphOutcomes(targetGraphs)
+        ).toResult
+      .onError: t =>
+        Logger[F]
+          .error(t)(s"Error calculating spectroscopy graph for input: $asterismRequest")
+
+  private def buildAsterismGraphRequest(
+    asterismRequest: AsterismSpectroscopyTimeRequest,
+    figures:         Option[SignificantFigures]
+  )(integrationTimes: AsterismIntegrationTimes): AsterismGraphRequest = {
+    val brightestTarget: TargetIntegrationTime = integrationTimes.value.focus
+    val selectedVariation: IntegrationTime     = brightestTarget.value.focus
+    val expTime: TimeSpan                      = selectedVariation.exposureTime
+    val expCount: PosInt                       = selectedVariation.exposures
+
+    AsterismGraphRequest(
+      asterismRequest.asterism,
+      GraphParameters(
+        asterismRequest.band,
+        asterismRequest.specMode,
+        asterismRequest.constraints,
+        expTime,
+        expCount,
+        asterismRequest.signalToNoiseAt
+      ),
+      figures
+    )
+  }
 
-  def versions[F[_]: Applicative: Logger](
-    environment: ExecutionEnvironment,
-    redis:       StringCommands[F, Array[Byte], Array[Byte]]
-  ): F[Result[ItcVersions]] =
-    Result(ItcVersions(version(environment).value, BuildInfo.ocslibHash.some)).pure[F]
+  def spectroscopyIntegrationTimeAndGraph[F[_]: MonadThrow: Logger: Parallel: Trace: Clock](
+    environment:     ExecutionEnvironment,
+    redis:           StringCommands[F, Array[Byte], Array[Byte]],
+    itc:             Itc[F]
+  )(
+    asterismRequest: AsterismSpectroscopyTimeRequest,
+    figures:         Option[SignificantFigures]
+  ): F[Result[SpectroscopyTimeAndGraphsResult]] =
+    ResultT(calculateSpectroscopyIntegrationTime(environment, redis, itc)(asterismRequest))
+      .flatMap: (specTimeResults: IntegrationTimeCalculationResult) =>
+        specTimeResults.targetTimes.partitionErrors
+          .bimap(
```

Comment:
So this makes itc return all results and all errors right? not failing all if one target fails.

Reply:
Exactly, all results are returned, whether they succeeded or not.

If there are errors, they are both in the result array in the `data` block and on an error array in the `errors` block. This way, the caller can decide whether to fail or not in case of partial errors.